### ISSUE TITLE
What to do about OpenGL

### ISSUE THREAD

**Issue by bradallred** [gemrb/gemrb#234]
see bradallred#112

the OpenGL backend will be unbuildable after #95 
Nobody still working on this project knows how to restore it to functionality, and it seems like we really won't have a need for it.

**Comment by MarcelHB** [comment]
I'll have two weeks of time by the weekend, so I can offer to look at it to bring it back to work at least.

I still have no big picture of the subviews code (I just did not look at it for two years), so any optimization ideas can be postponed.

**Comment by bradallred** [comment]
I’m not sure how much of an in depth view you would need, but here are a few important design changes that you definitely need. 

# Each window is its own render target
# SpriteCovers are now 8bit instead of 1bit
# GL should not be Subclassing the SDL2 driver. Subclassing the intermediate and abstract SDLVideo class is ok. 
# We have a global transparent render target for mouse and cursor/tooltips. I know GL could do this more efficiently, but that is baked into the WindowManager. 
# you will need to support at a minimum RGBA8888 and YV12 pixel formats for the buffers. RGB555 too if you feel like supporting 16bit mode (I wish we would just drop 16bit mode)
# The viewport is completely gone. This is handled directly by GameControl now. 
# I think we have a shader to deal with palettes, do you can ignore the versioning stuff I added for sdl2. 
# You can probably ignore the needsMask parameter on anything. This is due to the SDL2 limitation of being able to only blend 2 textures at once. OpenGL doesn’t have that limitation. 
# You don’t need to support any shape drawing code directly in OpenGL unless you want to. The only required primitive drawing requirements are DrawLine(s) and DrawPoint(s). We have software algorithms for doing everything else in terms of those. I encourage you to use those in place of the current GL implementation for now to keep all the tenderer s consistent code paths. If we need more performance later we can implement hardware accelerated versions. Keep in mind lynx and I know little about OpenGL so it’s best to use as much common code as possible IMO. 
# you should not try to handle BAMs. Return false for supporting those. 
# All the blitting APIs were simplified. Anchor points and the like are gone (viewport foo).
# I guess for now copy and paste the SDL2 driver code for setting up the window and handling input.

Look at SDLTextureVideoBuffer and SDLTextureSprite2D for hints. 

This might actually be easy for someone that knows what they are doing :)

**Comment by bradallred** [comment]
Sorry about the formatting. I posted from my phone and I can’t seem to edit... I’ll fix it later.

**Comment by MarcelHB** [comment]
Ok, my telco botched up badly. You probably won't hear much from me for a couple of days so I take your list and make it my offline project for the time. :tada:

**Comment by MarcelHB** [comment]
@bradallred Ok, I've made quite some progress, i.e. my main menu of IWD2 looks good and smooth. Since with the current OpenGL profiles, we don't have any real offscreen buffers, so I mainly draw to the screen, create a texture and then move it to its coordinates when doing `SwapBuffers`. So far so good.

One thing I'm currently try to solve: Say you click "Exit" in the main menu, then there is a window popping up. Now, for every `PushDrawingBuffer`, I clear/overwrite a rect of the screen large enough for the desired buffer dimensions, and then do a backup to texture.

Now, the dialog is drawn once and obviously does not do that again until required (`NeedsDraw()`), so my screen buffer remains blank and gets a blank backup to the texture, so I then paint transparent pixels instead of the window contents.

I suggest to to somehow extend the `VideoBuffer` to a method that allows indication of recycling, so I don't to a new backup?

**Comment by MarcelHB** [comment]
So, `VideoBuffer::Clear()` is simply expected to pixel-wise erase the buffer I guess, I don't want to mix that with _I do not overdraw the buffer right now._

**Comment by MarcelHB** [comment]
Ok, I tried the following thing successfully: For every call to `Draw()`, if anything below returns `true`, I let the subclass of `View` get a call that indicates that it can mark its buffer as untouched. My implementation is the only currently using it, but this solves the issue.

**Comment by bradallred** [comment]
If I'm understanding you correctly, it sounds like you opted to _not_ create a render texture for each window? As it sounds like you have noticed, that doesn't work due to overlapping windows. It sounds like perhaps worked around that, but I'm curious why not just follow the buffer-per-window pattern SDL uses which would fix that issue without changing the API. Windows aren't the only places where drawing buffers overlap BTW, the mouse cursor is another, I don't recall if any effects use them or not. Might just be the masking layer SDL 2 uses I'm thinking of.

yes, `VideoBuffer::Clear` is expected to set (as in completely wipeout anything drawn) all pixels in the buffer to transparent.

**Comment by bradallred** [comment]
I'm also not sure how you are getting away with drawing like that without completely changing the coordinates of draw calls. I thought all drawing was in window coordinates now. so are you not forcing yourself into another hack to workaround that?

**Comment by MarcelHB** [comment]
I do treat every `VideoBuffer` as a texture, I just use the screen as scratch place for everything until finally placing it where it blongs, since our OpenGL profiles don't provide -- or at least not without extensions -- any native framebuffer objects.

The coordinates are a nice thing to handle with matrices, we don't have to touch anything any more manually here. ;)

**Comment by bradallred** [comment]
ah, then its just my ignorance of OpenGL showing, carry on then sir!

I'm hoping the changes I made to how things are rendered make things more easy than difficult for you! I feel like a ton of hacks have been resolved now myself.

**Comment by bradallred** [comment]
Although, now I am curious how SDL2 itself deals with this. It uses opengl under the hood and it certainly isn't using the screen for any scratch drawing. 

Also, if you aren't aware there are several SDL OpenGL [APIs](https://wiki.libsdl.org/FrontPage?action=fullsearch&context=180&value=gl_&titlesearch=Titles) that might be useful to you.

**Comment by bradallred** [comment]
If I'm reading the SDL code right they are doing this:

```
if (SDL_GL_ExtensionSupported("GL_EXT_framebuffer_object")) {
        data->GL_EXT_framebuffer_object_supported = SDL_TRUE;
        data->glGenFramebuffersEXT = (PFNGLGENFRAMEBUFFERSEXTPROC)
            SDL_GL_GetProcAddress("glGenFramebuffersEXT");
        data->glDeleteFramebuffersEXT = (PFNGLDELETEFRAMEBUFFERSEXTPROC)
            SDL_GL_GetProcAddress("glDeleteFramebuffersEXT");
        data->glFramebufferTexture2DEXT = (PFNGLFRAMEBUFFERTEXTURE2DEXTPROC)
            SDL_GL_GetProcAddress("glFramebufferTexture2DEXT");
        data->glBindFramebufferEXT = (PFNGLBINDFRAMEBUFFEREXTPROC)
            SDL_GL_GetProcAddress("glBindFramebufferEXT");
        data->glCheckFramebufferStatusEXT = (PFNGLCHECKFRAMEBUFFERSTATUSEXTPROC)
            SDL_GL_GetProcAddress("glCheckFramebufferStatusEXT");
        renderer->info.flags |= SDL_RENDERER_TARGETTEXTURE;
    }
```

**Comment by MarcelHB** [comment]
Yes, that's the extension. There surely is a fallback, since my knowledge and ability to test mobile/OpenGL ES is limited, I rely on the code that is already in GemRB for now.

When it's repaired (YUV is still open, I'm currently chasing bugs) we can see if there is any benefit left.

**Comment by MarcelHB** [comment]
Ok, I vote for dropping OpenGL.

I've repaired the OpenGL backend under subviews to make the game work, leaving shaders and texturing as they are as far to consider some performance considerations, also w.r.t. what SDL2 brings from home.

The legacy version lived of some hacks that are easy to repair but also some terrible ones in the form of:

https://github.com/gemrb/gemrb/blob/aa38612084c4f239a257984c878ec4e6bb171e2f/gemrb/plugins/SDLVideo/SDL20GLVideo.cpp#L582

I didn't even know that this compiles in C++ without any kind of complaint (or there are some flags set somewhere?), but it strips the `const` away from the sprite!

Then you see that `GetTexture()` and some other calls lazily create an OpenGL texture, it even processes the video device mapping not earlier than that, i.e. lining up the color masks.

But let's say you are programmer that respects interfaces, or at least `const`s all the way down as they are specified in the `Video` and `Sprite2D` interfaces.

In our architecture, this brings you into some trouble: Calls to `glTexImage2D` is something you should not do more often than necessary, definitely not

1. over and over again for each frame and for all the plenty of sprites living on the current screen,
1. upfront for the whole population of sprites you create but don't want to bring to the screen any next frames, probably never.

Starting IWD2 into a save game in Targos had already ~25k sprites loaded. You can load textures frame by frame, but then I ended up at 15-20fps, wasting lots of work. You can also try to create a texture on creating the sprite immediately: This allocates an insane amount of memory via the user space driver, nothing a multi GB system could not handle but way too much for our purpose, and it also suffers from short framedrops when a new set of sprites is loading (e.g. new creatures and their animations).

Currently and without revamping some of the mentioned stuff, I don't see any code points of doing this loading better, for example batch uploads for things like cycling animation sets, a reuse of fitting allocated memory blocks. Just maybe keeping some stuff when copying sprites over resource holding. Generally, as you have seen there are extensions that ease coding such things, but then you need to test their availability upfront, and think of fallbacks or dropping platforms -- I'm not sure what we want to keep here, reading the README.

I mean given the scope this project visually has, I doubt there is much to win from doing all the mentioned work. Except SDL2 with subviews turns out to have some severe bottlenecks under normal conditions, then I really recommend to think of these tweaks.

But in the current state, it's more like a proof on concept or code you prefer not to maintain.

**Comment by bradallred** [comment]
`GLTextureSprite2D* glSprite = (GLTextureSprite2D*)spr;` is perfectly valid c++, no additional flags required, but in typical c++ it would be `GLTextureSprite2D* glSprite = const_cast<GLTextureSprite2D*>(spr);` instead. That sometimes has the potential to invoke Undefined Behavior (like trying to change data in read only memory), but in this case it's obviously going to work fine (as horrible as it is). It looks like we should be using `mutable` somewhere instead which would solve the `const` issue.

Part of the problem here also afflicts SDL2 pretty negatively, which is that BAMs are used to crate individual `Sprite2D` objects with their own textures instead of creating a single texture out of the BAM and sharing that between the sprites generated by that BAM. Correcting this is already on my wishlist, but I wouldn't think this alone would be a deal killer for OpenGL since SDL2 seems to cope ok. It is also non-trivial to do this cleanly. I was thinking about after subviews is merged and SDL 1.2 support is dropped.

The main bottlenecks (I've seen) with SDL2 on this branch is covered by #232 and #233 and both of those are pretty easy to solve. I imagine in an effect heavy fight with lots of palette swapping going on things might get noticeably slow.

If OpenGL were something somebody is interested in doing I would say use whatever extensions are required to do the job well. On Windows _everything_ added after 1.0 is an extension, so you will find it impossible to do most things on Windows if you are unwilling to use extensions :)

SDL2 driver already requires the `GL_EXT_framebuffer_object` and probably others. In fact, I'm pretty sure the shaders we have already require a version of OpenGL (GLSL) where these extensions are required and not really extensions at all.

**Comment by MarcelHB** [comment]
You're right, `mutable` looks like it could solve enough to just assign a GL texture handle here. Other than that, `const_cast` and its C style should not be used in this kind of scenario (interface gives you a `const` thing).

These are things I propose to do this job generally better:
- Reducing the number of sprites by two means: 
  1. Considering to create a large texture for one animation cycle. I don't expect the size to explode in general but we could then load it once and shift through the coordinates by frame.
  2. In the same process, already merge multipart stuff together into that texture, helmets, weapons, shadows. 
  3. This of course requires some invalidation for the case of changing equipment, applying spells  and similar.
- Have a texture and resource manager that can keep a number of textures alive in video memory, max size depending on the platform, eviction by the least number of active users. We can then serve hordes of goblins share the same animation cycle until the last one is gone.
- By raising OpenGL ES to 3.0, we still have iOS 7 and Android 4.3, and can use everything relevant of OpenGL 3.3 core. FBOs for free, and a way more comfortable GLSL as EGLSL 100/GLSL 120 have a bunch of limitations (such as no guarantee on efficient integers). I'm not sure if we'd still need both GLEW since SDL2 comes with methods that probably do the same thing.

This is a bit of work  and I haven't looked into font rendering yet.

**Comment by lynxlynxlynx** [comment]
We are the authors, so we can decide which gl version or extensions are mandatory. The fallback is always the software renderer of the main video driver, so there's no compatibility lost.

I'm pretty sure font's use i) already and we have a font atlas.

**Comment by bradallred** [comment]
> I suggest to simply focus on SDL2 and subviews for now, and then test where we hit the limits.
> 
> After that, somebody can focus on rewriting that from scratch with higher OGL profiles, and exploiting some of the things we mentioned here as general improvement for resources.

yup, I think that makes sense. General improvements to resource marshaling are required and when implemented SDL 2 alone might be "good enough". Battery saving on mobile devices might be the one thing a GL driver would be good for.

And yes, the original motivation behind the GL driver was to run on Android better. SDL 1.2 doesn't exist on mobile and our master SDL 2 driver is really just a gigantic hack that I threw together to get things working on iOS.
